File: backend/api/analytics.py
"""
Analytics API router.
Provides endpoint for reading analytics metrics.
"""

import logging
from fastapi import APIRouter, Depends, HTTPException
from middleware.auth import get_current_user
from services.analytics_service import AnalyticsService
from utils.supabase_client import supabase

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/analytics", tags=["analytics"])


@router.get("")
async def get_analytics(user: dict = Depends(get_current_user)):
    """Fetch reading analytics metrics for the authenticated user."""
    user_id = user.get("id") or user.get("sub")
    if not user_id:
        raise HTTPException(status_code=401, detail="Invalid token")

    try:
        metrics = AnalyticsService.calculate_dashboard_metrics(user_id, supabase)
        return metrics
    except Exception as e:
        logger.error("Failed to compute reading analytics dashboard: %s", e, exc_info=True)
        raise HTTPException(status_code=500, detail=f"Failed to compute reading analytics dashboard: {str(e)}")

File: backend_node/src/routes/analytics.ts
import { Router, Response } from 'express';
import { requireAuth, AuthenticatedRequest } from '../middleware/auth';
import { AnalyticsService } from '../services/analyticsService';
import { supabase } from '../config/supabase';

const router = Router();

function getUserId(req: AuthenticatedRequest): string {
  return (req.user?.id || req.user?.sub || '') as string;
}

router.get('/', requireAuth, async (req: AuthenticatedRequest, res: Response) => {
  const userId = getUserId(req);
  try {
    const metrics = await AnalyticsService.calculateDashboardMetrics(userId, supabase);
    return res.json(metrics);
  } catch (err: any) {
    console.error('[analytics] GET / error:', err);
    return res.status(500).json({ detail: err.message || String(err) });
  }
});

export default router;
